compiler/rustc_mir_transform/src/cross_crate_inline.rs

```diff
@@ -0,0 +1,125 @@
+use rustc_attr::InlineAttr;
+use rustc_hir::def::DefKind;
+use rustc_hir::def_id::LocalDefId;
+use rustc_middle::query::Providers;
+use rustc_middle::ty::TyCtxt;
+use rustc_session::config::OptLevel;
+
+pub fn provide(providers: &mut Providers) {
+    providers.cross_crate_inlinable = cross_crate_inlinable;
+}
+
+fn cross_crate_inlinable(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
+    // Obey source annotations first; this is important because it means we can use
+    // #[inline(never)] to force code generation.
+    match tcx.codegen_fn_attrs(def_id).inline {
+        InlineAttr::Never => return false,
+        InlineAttr::Hint | InlineAttr::Always => return true,
+        _ => {}
+    }
+
+    // This just reproduces the logic from Instance::requires_inline.
+    match tcx.def_kind(def_id) {
+        DefKind::Ctor(..) | DefKind::Closure => return true,
+        DefKind::Fn | DefKind::AssocFn => {}
+        _ => return false,
+    }
+
+    // Don't do any inference when incremental compilation is enabled; the additional inlining that
+    // inference permits also creates more work for small edits.
+    if tcx.sess.opts.incremental.is_some() {
+        return false;
+    }
+
+    // Don't do any inference unless optimizations are enabled.
+    if matches!(tcx.sess.opts.optimize, OptLevel::No | OptLevel::Default) {
+        return false;
+    }
+
+    if !tcx.is_mir_available(def_id) {
+        return false;
+    }
+
+    let mir = tcx.optimized_mir(def_id);
+    let mut checker =
+        CostChecker { tcx, callee_body: mir, calls: 0, statements: 0, landing_pads: 0, resumes: 0 };
+    checker.visit_body(mir);
+    if checker.calls == 0
+        && checker.resumes == 0
+        && checker.landing_pads == 0
+        && checker.statements <= 100
+    {
+        /*
+        let span = tcx.def_span(def_id);
+        tcx.sess.emit_warning(crate::errors::SuggestAddingInline {
+            place: span,
+            suggest_inline: span.with_hi(span.lo()),
+            statements: checker.statements,
+        });
+        */
+        true
+    } else {
+        false
+    }
+}
+
+use rustc_middle::mir::visit::Visitor;
+use rustc_middle::mir::*;
+
+struct CostChecker<'b, 'tcx> {
```

Comment:
On the first point: https://github.com/rust-lang/rust/pull/107009/commits/7710d8011e93e4738075f7f30193868a82371e32
On the second point: if a body passes your criterion, then its cost is <= 500. That's much higher than the current inlining cost threshold (125 for functions with <= 3 blocks, so closer to 25 statements). This is one of the reasons I propose reusing the same cost model: to compare them easily.
Should the cost for automatic inlining be commensurate to the inlining threshold? How much higher to account for subtitutions? Or just reuse the same threshold for simplicity? Or much higher as you propose, because LLVM has a different analysis.

Reply:
I'm very confused. when I said the same cost model doesn't make sense, what I meant is the relationship between the cost of a statement and the cost of a call and all the other constructs in MIR. The evidence from the perf runs in this PR very strongly indicate that the correct cost model here imposes a vastly greater penalty on calls.

Reply:
I've updated the PR description to reflect that I do not think the heuristic implemented here is ideal, but simply that it produces very good results in the perf suite right now so I'd prefer to land this and continue experimenting in another PR.